sway/config/output.c

```diff
@@ -0,0 +1,181 @@
+		return NULL;
+	}
+	oc->enabled = -1;
+	oc->width = oc->height -1;
+	oc->refresh_rate = -1;
+	oc->x = oc->y = -1;
+	oc->scale = -1;
+	oc->transform = -1;
+	return oc;
+}
+
+void merge_output_config(struct output_config *dst, struct output_config *src) {
+	if (src->name) {
+		free(dst->name);
+		dst->name = strdup(src->name);
+	}
+	if (src->enabled != -1) {
+		dst->enabled = src->enabled;
+	}
+	if (src->width != -1) {
+		dst->width = src->width;
+	}
+	if (src->height != -1) {
+		dst->height = src->height;
+	}
+	if (src->x != -1) {
+		dst->x = src->x;
+	}
+	if (src->y != -1) {
+		dst->y = src->y;
+	}
+	if (src->scale != -1) {
+		dst->scale = src->scale;
+	}
+	if (src->refresh_rate != -1) {
+		dst->refresh_rate = src->refresh_rate;
+	}
+	if (src->transform != -1) {
+		dst->transform = src->transform;
+	}
+	if (src->background) {
+		free(dst->background);
+		dst->background = strdup(src->background);
+	}
+	if (src->background_option) {
+		free(dst->background_option);
+		dst->background_option = strdup(src->background_option);
+	}
+}
+
+static void set_mode(struct wlr_output *output, int width, int height,
+		float refresh_rate) {
+	struct wlr_output_mode *mode, *best = NULL;
+	int mhz = (int)(refresh_rate * 1000);
+	wl_list_for_each(mode, &output->modes, link) {
+		if (mode->width == width && mode->height == height) {
+			if (mode->refresh == mhz) {
+				best = mode;
+				break;
+			}
+			best = mode;
+		}
+	}
+	if (!best) {
+		sway_log(L_ERROR, "Configured mode for %s not available", output->name);
+	} else {
+		sway_log(L_DEBUG, "Assigning configured mode to %s", output->name);
+		wlr_output_set_mode(output, best);
+	}
+}
+
+void apply_output_config(struct output_config *oc, swayc_t *output) {
+	assert(output->type == C_OUTPUT);
+
+	struct wlr_output *wlr_output = output->sway_output->wlr_output;
+	if (oc && oc->enabled == 0) {
+		wlr_output_layout_remove(root_container.output_layout, wlr_output);
+		destroy_output(output);
+		return;
+	}
+
+	if (oc && oc->width > 0 && oc->height > 0) {
+		set_mode(wlr_output, oc->width, oc->height, oc->refresh_rate);
+	}
+	if (oc && oc->scale > 0) {
+		sway_log(L_DEBUG, "Set %s scale to %d", oc->name, oc->scale);
+		wlr_output_set_scale(wlr_output, oc->scale);
+	}
+	if (oc && oc->transform >= 0) {
+		sway_log(L_DEBUG, "Set %s transform to %d", oc->name, oc->transform);
+		wlr_output_transform(wlr_output, oc->transform);
+	}
+
+	// Find position for it
+	if (oc && (oc->x != -1 || oc->y != -1)) {
+		sway_log(L_DEBUG, "Set %s position to %d, %d", oc->name, oc->x, oc->y);
+		wlr_output_layout_add(root_container.output_layout, wlr_output, oc->x,
+			oc->y);
+	} else {
+		wlr_output_layout_add_auto(root_container.output_layout, wlr_output);
+	}
+	struct wlr_box *output_layout_box =
+		wlr_output_layout_get_box(root_container.output_layout, wlr_output);
+	output->x = output_layout_box->x;
+	output->y = output_layout_box->y;
+	output->width = output_layout_box->width;
+	output->height = output_layout_box->height;
```

Comment:
The output container could listen to the output layout change event and automatically keep these in sync.